samples/boards/nrf/battery/src/battery.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2018-2019 Peter Bigot Consulting, LLC
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <zephyr.h>
+#include <init.h>
+#include <drivers/gpio.h>
+#include <drivers/adc.h>
+#include <drivers/sensor.h>
+#include <logging/log.h>
+
+#include "battery.h"
+
+LOG_MODULE_REGISTER(BATTERY, CONFIG_ADC_LOG_LEVEL);
+
+struct io_channel_config {
+	const char *label;
+	u8_t channel;
+};
+
+struct gpio_channel_config {
+	const char *label;
+	u8_t pin;
+	u8_t flags;
+};
+
+struct divider_config {
+	const struct io_channel_config io_channel;
+	const struct gpio_channel_config power_gpios;
+	const u32_t output_ohm;
+	const u32_t full_ohm;
+};
+
+static const struct divider_config divider_config = {
+	.io_channel = DT_VOLTAGE_DIVIDER_VBATT_IO_CHANNELS,
+#ifdef DT_VOLTAGE_DIVIDER_VBATT_POWER_GPIOS
+	.power_gpios = DT_VOLTAGE_DIVIDER_VBATT_POWER_GPIOS,
+#endif
+	.output_ohm = DT_VOLTAGE_DIVIDER_VBATT_OUTPUT_OHMS,
+	.full_ohm = DT_VOLTAGE_DIVIDER_VBATT_FULL_OHMS,
+};
+
+struct divider_data {
+	struct device *adc;
+	struct device *gpio;
+	struct adc_channel_cfg adc_cfg;
+	struct adc_sequence adc_seq;
+	s16_t raw;
+};
+static struct divider_data divider_data;
+
+static int divider_setup(void)
+{
+	const struct divider_config *cfg = &divider_config;
+	const struct io_channel_config *iocp = &cfg->io_channel;
+	const struct gpio_channel_config *gcp = &cfg->power_gpios;
+	struct divider_data *ddp = &divider_data;
+	struct adc_sequence *asp = &ddp->adc_seq;
+	struct adc_channel_cfg *accp = &ddp->adc_cfg;
+	int rc;
+
+	ddp->adc = device_get_binding(iocp->label);
+	if (ddp->adc == NULL) {
+		LOG_ERR("Failed to get ADC %s", iocp->label);
+		return -ENOENT;
+	}
+
+	if (gcp->label) {
+		ddp->gpio = device_get_binding(gcp->label);
+		if (ddp->gpio == NULL) {
+			LOG_ERR("Failed to get GPIO %s", gcp->label);
+			return -ENOENT;
+		}
+		rc = gpio_pin_write(ddp->gpio, gcp->pin, 0);
+		if (rc == 0) {
+			rc = gpio_pin_configure(ddp->gpio, gcp->pin,
+						gcp->flags | GPIO_DIR_OUT);
+		}
+		if (rc != 0) {
+			LOG_ERR("Failed to control feed %s.%u: %d",
+				gcp->label, gcp->pin, rc);
+			return rc;
+		}
+	}
+
+	*asp = (struct adc_sequence){
+		.channels = BIT(0),
+		.buffer = &ddp->raw,
+		.buffer_size = sizeof(ddp->raw),
+		.calibrate = true,
+	};
+
+#ifdef CONFIG_ADC_NRFX_SAADC
+	*accp = (struct adc_channel_cfg){
+		.gain = ADC_GAIN_1_6,
```

Comment:
on thingy voltage divider ensures that result will be less than 0.6V so `ADC_GAIN_1_6` could be used. Additionally when i added `.oversampling = 4` to adc sequence i got stable results:
```
[0:02:09.362]: 3630 mV; 2500 pptt
[0:02:14.337]: 3630 mV; 2500 pptt
[0:02:19.310]: 3630 mV; 2500 pptt
[0:02:24.284]: 3630 mV; 2500 pptt
[0:02:29.259]: 3630 mV; 2500 pptt
[0:02:34.233]: 3621 mV; 2289 pptt
[0:02:39.208]: 3630 mV; 2500 pptt
[0:02:44.181]: 3630 mV; 2500 pptt
[0:02:49.155]: 3630 mV; 2500 pptt
[0:02:54.129]: 3630 mV; 2500 pptt
```

Reply:
Updated to use oversampling and fix some CI whinage.

Did you mean `ADC_GAIN_1` could be used?  Yes, for nrf52_pca20020 where the voltage divider scales by 0.10714 (assuming Vbatt maxes at 4.2 V with a LiPo the voltage presented to the ADC is at most 450 mV).  The divisor used on particle xenon (which this sample also supports) scales by 0.72264 (input voltage up to ~ 3.0 V).  For that target only 1/6 or 1/5 gain is sure to work, and 1/5 is borderline.

Reply:
could we have some method of getting optimal gain? e.g. what if voltage divider gets `max-full-voltage` field from which max voltage on divider could be derived and then optimal gain (based on reference).

Reply:
otherwise this sample on thingy52 gives impression that something is wrong or that measurement is pointless.

Reply:
> otherwise this sample on thingy52 gives impression that something is wrong or that measurement is pointless.

Can you explain that more?  I don't see what suggests the measurement is wrong.  The sample is simply designed to support the ADC being given the highest supported voltage that could be presented to the chip without burning it out: 3.6 V => 600 mV with gain 1/6.

Reply:
>  I don't see what suggests the measurement is wrong

Output like one below which shows battery level from 14% to 33% which rather tells that battery level cannot be measured reliably.
```
[0:07:02.564]: 3612 mV; 2078 pptt
[0:07:07.535]: 3621 mV; 2289 pptt
[0:07:12.505]: 3630 mV; 2500 pptt
[0:07:17.475]: 3584 mV; 1421 pptt
[0:07:22.444]: 3630 mV; 2500 pptt
[0:07:27.413]: 3621 mV; 2289 pptt
[0:07:32.384]: 3640 mV; 2734 pptt
[0:07:37.356]: 3640 mV; 2734 pptt
[0:07:42.328]: 3630 mV; 2500 pptt
[0:07:47.300]: 3630 mV; 2500 pptt
[0:07:52.271]: 3668 mV; 3390 pptt
[0:07:57.243]: 3640 mV; 2734 pptt
[0:08:02.215]: 3593 mV; 1632 pptt
```

As you said, because of divider maximal voltage on pin is ~450mV so having gain_1 would be optimal to measure it with 600mV reference and having 1/6 leads to fluctuations like above.

Reply:
OK, I modified the example to special case Thingy:52.  In the general case you can't infer the correct gain to reduce, because you don't know what the battery output voltage is.